### PR TITLE
WIP: rustup

### LINKED ISSUE #1371: does not compile with rustc 1.15.0-nightly (1c448574b 2016-11-28)
A ton of stuff seems to have been renamed. Errors can be seen here: https://travis-ci.org/durka/guard/jobs/180149918#L253

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #1371

going to bed. Didn't get this fixed all the way. I think the `vec!` macro changed, so `higher::vec_macro` is borked (but not in the part that I changed I think)

### CODE REVIEW

**Comment by mcarton** [comment]
@oli-obk did you find the responsible PR from rustc? That usually really helps.

**Comment by oli-obk** [comment]
https://github.com/rust-lang/rust/pull/37676 that was one of the changes, but there are multiple things going on at once I think

**Comment by oli-obk** [comment]
hashes of paths are wrong... for `copies.rs`

not sure how to fix `enum_glob_use`

**Comment by mcarton** [comment]
> hashes of paths are wrong... for `copies.rs`

As long as equality works and `a == b ⇒ hash(a) == hash(b)` it's fine. You don't need perfect hashes.

**Comment by oli-obk** [comment]
ok... we're down to equal for loops not comparing equal

**Comment by oli-obk** [comment]
If someone could pick this up, that would be great. I'm not gonna be near a pc until sunday.

All that is left is to dump the failing test (copy just the failing expression to a new file and dump the ast) and figure out what part of a desugared for loop fails to compare

File: clippy_lints/src/utils/hir.rs [hunk 1]
```diff
@@ -178,40 +172,23 @@ impl<'a, 'tcx: 'a> SpanlessEq<'a, 'tcx> {
         }
     }
 
-    fn eq_path(&self, left: &Path, right: &Path) -> bool {
-        left.global == right.global &&
-        over(&left.segments, &right.segments, |l, r| self.eq_path_segment(l, r))
-    }
+    fn eq_qpath(&self, left: &QPath, lid: NodeId, right: &QPath, rid: NodeId) -> bool {
+        let l = self.cx.tcx.tables().qpath_def(left, lid);
+        let r = self.cx.tcx.tables().qpath_def(right, rid);
```
Comment (mcarton) [inline clippy_lints/src/utils/hir.rs]: Now `foo::<u8>` and `foo::<u32>` compare equal, this is wrong.

File: clippy_lints/src/utils/hir.rs [hunk 2]
```diff
@@ -100,11 +101,11 @@ impl<'a, 'tcx: 'a> SpanlessEq<'a, 'tcx> {
                 self.eq_expr(lc, rc) && self.eq_block(lt, rt) && both(le, re, |l, r| self.eq_expr(l, r))
             }
             (&ExprLit(ref l), &ExprLit(ref r)) => l.node == r.node,
-            (&ExprLoop(ref lb, ref ll, ref lls), &ExprLoop(ref rb, ref rl, ref rls)) => {
```
Comment (mcarton) [inline clippy_lints/src/utils/hir.rs]: Why did you remove the loop source (`{l,r}ls`)? It's just a variant-less `Copy` enum.

File: clippy_lints/src/utils/hir.rs [hunk 3]
```diff
@@ -100,11 +101,11 @@ impl<'a, 'tcx: 'a> SpanlessEq<'a, 'tcx> {
...
-                self.eq_block(lb, rb) && both(ll, rl, |l, r| l.node.as_str() == r.node.as_str()) && lls == rls
+            (&ExprLoop(ref lb, ref ll, _), &ExprLoop(ref rb, ref rl, _)) => {
+                self.eq_block(lb, rb) && both(ll, rl, |l, r| l.node.as_str() == r.node.as_str())
             }
-            (&ExprMatch(ref le, ref la, ref ls), &ExprMatch(ref re, ref ra, ref rs)) => {
-                ls == rs && self.eq_expr(le, re) &&
+            (&ExprMatch(ref le, ref la, _), &ExprMatch(ref re, ref ra, _)) => {
```
Comment (mcarton) [inline clippy_lints/src/utils/hir.rs]: Ditto.

File: clippy_lints/src/utils/hir.rs [hunk 4]
```diff
@@ -409,16 +380,15 @@ impl<'a, 'tcx: 'a> SpanlessHash<'a, 'tcx> {
...
             }
-            ExprLoop(ref b, ref i, ref j) => {
+            ExprLoop(ref b, ref i, _) => {
                 let c: fn(_, _, _) -> _ = ExprLoop;
                 c.hash(&mut self.s);
                 self.hash_block(b);
                 if let Some(i) = *i {
                     self.hash_name(&i.node);
                 }
-                j.hash(&mut self.s);
```
Comment (mcarton) [inline clippy_lints/src/utils/hir.rs]: Ditto.

**Comment by mcarton** [comment]
This works locally on `rustc 1.15.0-nightly (1c448574b 2016-11-28)` now but it looks like it need more updates for `rustc 1.15.0-nightly (908dba0c9 2016-12-01)` :sweat:

File: clippy_lints/src/enum_glob_use.rs [hunk 5]
```diff
@@ -48,24 +46,12 @@ impl EnumGlobUse {
...
-                                span_lint(cx, ENUM_GLOB_USE, item.span, "don't use glob imports for enum variants");
-                            }
-                        }
-                    }
-                }
+        if let ItemUse(ref path, UseKind::Glob) = item.node {
+            // FIXME: ask jseyfried why the qpath.def for `use std::cmp::Ordering::*;`
+            // extracted through `ItemUse(ref qpath, UseKind::Glob)` is a `Mod` and not an `Enum`
+            //if let Def::Enum(_) = path.def {
+            if path.segments.last().and_then(|seg| seg.name.as_str().chars().next()).map_or(false, char::is_uppercase) {
```
Comment (mcarton) [inline clippy_lints/src/enum_glob_use.rs]: Do you had an answer for that?

File: clippy_lints/src/utils/mod.rs [hunk 6]
```diff
@@ -94,7 +95,16 @@ pub fn differing_macro_contexts(lhs: Span, rhs: Span) -> bool {
...
+    cx.sess().codemap().with_expn_info(span.expn_id, |info| match info {
+        Some(info) => {
+            match info.callee.format {
+                // don't treat range expressions desugared to structs as "in_macro"
+                ExpnFormat::CompilerDesugaring(name) => name != "...",
+                _ => true,
+            }
+        },
+        None => false,
+    })
```
Comment (mcarton) [inline clippy_lints/src/utils/mod.rs]: Why did you have to change this? IMO we don't have enough `in_macro` tests.

File: tests/compile-fail/dlist.rs [hunk 7]
```diff
@@ -10,5 +10,5 @@ pub fn test(_: LinkedList<u8>) {  //~ ERROR I see you're using a LinkedList!
 }
 
 fn main(){
-    test(LinkedList::new());
+    test(LinkedList::new()); //~ ERROR I see you're using a LinkedList!
```
Comment (mcarton) [inline tests/compile-fail/dlist.rs]: I thought this wasn't here on purpose: if `test` takes a `LinkedList`, the user has no choice but to build one to call the function. Hence the function should be warned about, not its call.

**Comment by mcarton** [comment]
The new update for `rustc 1.15.0-nightly (908dba0c9 2016-12-01)` is quite big, I will finish that tomorrow.

**Comment by jseyfried** [comment]
@oli-obk 
> Why is the `qpath.def` for `use std::cmp::Ordering::*;` a `Mod` and not an `Enum`?

Because [this code](https://github.com/rust-lang/rust/blob/c80c31a502c838f9ec06f1003d7c61cf9de9d551/src/librustc_resolve/resolve_imports.rs#L699) is incorrect -- it should record `module.def().unwrap()`, not `Def::Mod(module.def_id().unwrap())`.
This is a very minor change -- I'll fix in my next PR (probably tomorrow).

File: clippy_lints/src/utils/mod.rs (hunk 6 above)
Comment (oli-obk) [inline clippy_lints/src/utils/mod.rs]: Because without this, we can't do anything useful with ranges in hir, since they all have that expansion span

File: clippy_lints/src/utils/hir.rs (hunk 4 above)
Comment (oli-obk) [inline clippy_lints/src/utils/hir.rs]: Because they don't really add any value... If a desugaring is the same as a manual write then it is the same

**Comment by oli-obk** [comment]
> The new update for rustc 1.15.0-nightly (908dba0c9 2016-12-01) is quite big too, I will finish that tomorrow.

we need to patch rustc first. https://github.com/rust-lang/rust/pull/37918 broke all lints that forward to an inner `Visitor`, since `LateContext`'s methods do not have lifetime annotations binding the given ast/hir node's lifetime to the `TyCtxt`'s. A hacky workaround in clippy would be to do an intermediate call to `tcx.map.expect_*` which again yields a `tcx` lifetime, but I think we'd save us many headaches by simply fixing rustc.

**Comment by oli-obk** [comment]
I have a rustc pr ready but didn't get clippy ported all the way. Will finish tomorrow

**Comment by oli-obk** [comment]
got it to work with https://github.com/rust-lang/rust/pull/38191

**Comment by oli-obk** [comment]
rust PR has been merged, now we wait for a new nightly to survive the buildbots: http://rusty-dash.com/nightlies

**Comment by Arnavion** [comment]
https://github.com/rust-lang/rust/pull/38097/ broke this a little more. `rustc::ty::FnSig` no longer has fields called `inputs` and `output`

**Comment by mcarton** [comment]
Thanks @Arnavion!
Now we just have to wait for a new nightly to be published, but [it's been a few days since there hasn't been any](http://rusty-dash.com/nightlies).

**Comment by Wabuo** [comment]
There is a nightly release now ...

**Comment by oli-obk** [comment]
Travis doesn't have it yet

**Comment by mcarton** [comment]
:tada: Published :tada:
